sound/drivers/pimidi.c

```diff
@@ -0,0 +1,1119 @@
+
+	if (instance->stopping) {
+		printd(instance, "DRDY interrupt, but stopping, ignoring...");
+		return IRQ_HANDLED;
+	}
+
+	printd(instance, "DRDY interrupt, masking");
+	disable_irq_nosync(irq);
+
+	printd(instance, "Queue work due to DRDY interrupt");
+	queue_work(pimidi_global.work_queue, &instance->drdy_handler);
+
+	return IRQ_HANDLED;
+}
+
+static void pimidi_proc_stat_show(struct snd_info_entry *entry, struct snd_info_buffer *buffer)
+{
+	const unsigned int *d = entry->private_data;
+
+	snd_iprintf(buffer, "%u\n", *d);
+}
+
+static void pimidi_proc_serial_show(struct snd_info_entry *entry, struct snd_info_buffer *buffer)
+{
+	struct pimidi_instance *instance = entry->private_data;
+
+	snd_iprintf(buffer, "%s\n", instance->serial);
+}
+
+static void pimidi_proc_version_show(struct snd_info_entry *entry, struct snd_info_buffer *buffer)
+{
+	struct pimidi_instance *instance = entry->private_data;
+
+	snd_iprintf(buffer, "%u.%u.%u\n", instance->version.major, instance->version.minor,
+		    instance->version.build);
+}
+
+static void pimidi_proc_hwrev_show(struct snd_info_entry *entry, struct snd_info_buffer *buffer)
+{
+	struct pimidi_instance *instance = entry->private_data;
+
+	snd_iprintf(buffer, "%u\n", instance->version.hwrev);
+}
+
+static int pimidi_i2c_probe(struct i2c_client *client)
+{
+	struct snd_card *card = NULL;
+	int err, d, i;
+
+	bool irq_initialized = false;
+
+	d = client->addr - 0x20;
+
+	if (d < 0 || d >= 8) {
+		printe_g("Unexpected device address: %d", client->addr);
+		err = -EINVAL;
+		goto finalize;
+	}
+
+	err = snd_card_new(&client->dev, SNDRV_DEFAULT_IDX1, SNDRV_DEFAULT_STR1, THIS_MODULE,
+			   sizeof(struct pimidi_instance), &card);
+
+	if (err) {
+		printe_g("snd_card_new failed: %d", err);
+		return err;
+	}
+
+	struct pimidi_instance *instance = (struct pimidi_instance *)card->private_data;
+
+	instance->i2c_client = client;
+	instance->d = d;
+
+	struct snd_rawmidi *rawmidi;
+
+	err = snd_rawmidi_new(card, card->shortname, 0, 2, 2, &rawmidi);
+	if (err < 0) {
+		printe(instance, "snd_rawmidi_new failed: %d", err);
+		goto finalize;
+	}
+
+	instance->rawmidi = rawmidi;
+	strscpy(rawmidi->name, "pimidi", sizeof(rawmidi->name));
+
+	rawmidi->info_flags =
+		SNDRV_RAWMIDI_INFO_OUTPUT | SNDRV_RAWMIDI_INFO_INPUT | SNDRV_RAWMIDI_INFO_DUPLEX;
+	rawmidi->private_data = instance;
+	rawmidi->ops = &pimidi_midi_ops;
+
+	snd_rawmidi_set_ops(rawmidi, SNDRV_RAWMIDI_STREAM_OUTPUT, &pimidi_midi_output_ops);
+	snd_rawmidi_set_ops(rawmidi, SNDRV_RAWMIDI_STREAM_INPUT, &pimidi_midi_input_ops);
+
+	instance->data_ready_gpio = devm_gpiod_get(&client->dev, "data-ready", GPIOD_OUT_HIGH);
+	if (IS_ERR(instance->data_ready_gpio)) {
+		err = PTR_ERR(instance->data_ready_gpio);
+		printe(instance, "devm_gpiod_get failed: %d", err);
+		goto finalize;
+	}
+
+	err = pimidi_register(instance);
+	if (err < 0) {
+		printe(instance, "pimidi_register failed: %d", err);
+		goto finalize;
+	}
+
+	pimidi_perform_reset();
```

Comment:
Can you explain to me why you go to the trouble of resetting the other devices when each instance is probed? I understand that it is a shared reset line, but why can't you just reset when the first instance is probed? Without this, the driver would be much simpler.

Reply:
This supports the `sudo dtoverlay pimidi sel=x` usage when the devices must be loaded dynamically. As the reset is shared, each device must be re-initialized.

Reply:
One of us isn't understanding. If the reset is shared, isn't all the hardware reset by the first probe?

Reply:
Each device must be initialized by a special 'handshake' via the DRDY GPIO. When a new device is added dynamically to the device tree, the kernel module can't know what is the current state of the new device, therefore, it must reset and do the 'handshake' procedure to all of them all over again, to ensure the state is consistent.

Reply:
Keep in mind that when initial device is added via dtoverlay, the kernel module only knows about 1 of the DRDY pins, and a probe gets executed immediately. When a 2nd device gets added, it now knows about 2 DRDY pins, so it must repeat the procedure, as the new device must be reset, and as the reset is a shared line, it must re-initialize all the devices it knew of up till now.